File: api/transcript.py
from http.server import BaseHTTPRequestHandler
from youtube_transcript_api import YouTubeTranscriptApi
import json
import re

def extract_video_id(url):
    patterns = [
        r'(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([^&\n?#]+)',
        r'^([a-zA-Z0-9_-]{11})$'
    ]
    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            return match.group(1)
    return None

def format_timestamp(seconds):
    minutes = int(seconds // 60)
    secs = int(seconds % 60)
    return f"{minutes:02d}:{secs:02d}"

def format_to_markdown(transcript, video_url, video_id):
    markdown = f"# YouTube Video Transcript\n\n"
    markdown += f"**Video URL:** {video_url}\n\n"
    markdown += f"**Video ID:** {video_id}\n\n"
    markdown += "---\n\n"
    markdown += "## Transcript\n\n"
    for item in transcript:
        timestamp = format_timestamp(item.start)
        markdown += f"**[{timestamp}]** {item.text}\n\n"
    return markdown

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)

        try:
            data = json.loads(post_data.decode('utf-8'))
            url = data.get('url', '').strip()

            if not url:
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'YouTube URL is required'}).encode())
                return

            video_id = extract_video_id(url)

            if not video_id:
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Invalid YouTube URL'}).encode())
                return

            ytt = YouTubeTranscriptApi()
            transcript = None

            try:
                transcript = ytt.fetch(video_id)
            except Exception:
                try:
                    available = ytt.list(video_id)
                    if available:
                        for t in available:
                            try:
                                transcript = t.fetch()
                                break
                            except:
                                continue
                except:
                    pass

            if not transcript:
                self.send_response(404)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'No transcript available for this video'}).encode())
                return

            markdown = format_to_markdown(transcript, url, video_id)

            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'success': True,
                'videoId': video_id,
                'markdown': markdown,
                'filename': f'transcript-{video_id}.md'
            }).encode())

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

File: api/transcript.js
const https = require('https');

function extractVideoId(url) {
  const patterns = [
    /(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([^&\n?#]+)/,
    /^([a-zA-Z0-9_-]{11})$/
  ];
  for (const pattern of patterns) {
    const match = url.match(pattern);
    if (match) return match[1];
  }
  return null;
}

function fetchUrl(url) {
  return new Promise((resolve, reject) => {
    https.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept-Language': 'en-US,en;q=0.9'
      }
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

function parseCaptions(xml) {
  const captions = [];
  const regex = /<text start="([^"]*)" dur="([^"]*)"[^>]*>([\s\S]*?)<\/text>/g;
  let match;
  while ((match = regex.exec(xml)) !== null) {
    const text = match[3]
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'")
      .replace(/\n/g, ' ')
      .trim();
    if (text) {
      captions.push({
        start: parseFloat(match[1]),
        text
      });
    }
  }
  return captions;
}

async function getCaptionsUrl(videoId) {
  const pageUrl = `https://www.youtube.com/watch?v=${videoId}`;
  const html = await fetchUrl(pageUrl);

  const timedtextMatch = html.match(/timedtext\?v=[^"]+lang=en[^"]*/);
  if (timedtextMatch) {
    let url = timedtextMatch[0].replace(/\\u0026/g, '&');
    return `https://www.youtube.com/api/${url}`;
  }

  const anyTimedtext = html.match(/timedtext\?v=[^"]+/);
  if (anyTimedtext) {
    let url = anyTimedtext[0].replace(/\\u0026/g, '&');
    return `https://www.youtube.com/api/${url}`;
  }

  return null;
}

function formatTimestamp(seconds) {
  const mins = Math.floor(seconds / 60);
  const secs = Math.floor(seconds % 60);
  return `${mins.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
}

function formatToMarkdown(transcript, videoUrl, videoId) {
  let md = `# YouTube Video Transcript\n\n`;
  md += `**Video URL:** ${videoUrl}\n\n`;
  md += `**Video ID:** ${videoId}\n\n`;
  md += `---\n\n## Transcript\n\n`;
  transcript.forEach(item => {
    md += `**[${formatTimestamp(item.start)}]** ${item.text}\n\n`;
  });
  return md;
}

module.exports = async function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const { url } = req.body;

    if (!url) {
      return res.status(400).json({ error: 'YouTube URL is required' });
    }

    const videoId = extractVideoId(url);
    if (!videoId) {
      return res.status(400).json({ error: 'Invalid YouTube URL' });
    }

    const captionsUrl = await getCaptionsUrl(videoId);
    if (!captionsUrl) {
      return res.status(404).json({ error: 'No transcript available for this video' });
    }

    const captionsXml = await fetchUrl(captionsUrl);
    const transcript = parseCaptions(captionsXml);

    if (!transcript || transcript.length === 0) {
      return res.status(404).json({ error: 'No transcript available for this video' });
    }

    const markdown = formatToMarkdown(transcript, url, videoId);

    return res.status(200).json({
      success: true,
      videoId,
      markdown,
      filename: `transcript-${videoId}.md`
    });

  } catch (error) {
    console.error('Error:', error);
    return res.status(500).json({ error: error.message || 'Failed to fetch transcript' });
  }
};
